node/subsystem-util/src/nesting_sender.rs

```diff
@@ -0,0 +1,148 @@
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Hierarchical messages.
+//!
+//! Writing concurrent and even multithreaded by default is inconvenient and slow: No references
+//! hence lots of needless cloning and data duplication, locks, mutexes, ... We should reach
+//! for concurrency and parallelism when there is an actual need, not just because we can and it is
+//! reasonably safe in Rust.
+//!
+//! I very much agree with many points in this blog post for example:
+//!
+//! https://maciej.codes/2022-06-09-local-async.html
+//!
+//! Another very good post by Pierre (Tomaka):
+//!
+//! https://tomaka.medium.com/a-look-back-at-asynchronous-rust-d54d63934a1c
+//!
+//! This module helps with this in part. It does not break the multithreaded by default approach,
+//! but it breaks the `spawn everything` approach. So once you `spawn` you will still be
+//! multithreaded by default, despite that for most tasks we spawn (which just wait for network or some
+//! message to arrive), that is very much pointless and needless overhead. You will just spawn less in
+//! the first place.
+//!
+//! By default your code is single threaded, except when actually needed:
+//!		- need to wait for long running synchronous IO (a threaded runtime is actually useful here)
+//!		- need to wait for some async event (message to arrive)
+//!		- need to do some hefty CPU bound processing (a thread is required here as well)
+//!
+//! and it is not acceptable to block the main task for waiting for the result, because we actually
+//! really have other things to do or at least need to stay responsive just in case.
+//!
+//! With the types and traits in this module you can achieve exactly that: You write modules which
+//! just execute logic and can call into the functions of other modules - yes we are calling normal
+//! functions. For the case a module you are calling into requires an occasional background task,
+//! you provide it with a `NestingSender<M, ChildModuleMessage>` that it can pass to any spawned tasks.
+//!
+//! This way you don't have to spawn a task for each module just for it to be able to handle
+//! asynchronous events. The module relies on the using/enclosing code/module to forward it any
+//! asynchronous messages in a structured way.
+//!
+//! What makes this architecture nice is the separation of concerns - at the top you only have to
```

Comment:
nit: Just some quick feedback here, and maybe it's just me. But not having looked at any code or examples yet, I was a little confused where the nesting comes in. I think I get it after a couple re-reads, but could be more clear.

Reply:
Added example, hope that helps.